internal/mq/msgstream/mqwrapper/nmq/nmq_consumer.go

```diff
@@ -0,0 +1,161 @@
+	natsChan  chan *nats.Msg
+	msgChan   chan mqwrapper.Message
+	closeChan chan struct{}
+	once      sync.Once
+	skip      bool
+}
+
+// Subscription returns the subscription name of this consumer
+func (nc *Consumer) Subscription() string {
+	return nc.groupName
+}
+
+// Chan returns a channel to read messages from natsmq
+func (nc *Consumer) Chan() <-chan mqwrapper.Message {
+	if err := nc.closed(); err != nil {
+		panic(err)
+	}
+
+	if nc.sub == nil {
+		log.Error("accessing Chan of an uninitialized subscription.", zap.String("topic", nc.topic), zap.String("groupName", nc.groupName))
+		panic("failed to chan a consumer without assign")
+	}
+	if nc.msgChan == nil {
+		nc.once.Do(func() {
+			nc.msgChan = make(chan mqwrapper.Message, 256)
+			go func() {
+				for {
+					select {
+					case msg := <-nc.natsChan:
+						if nc.skip {
+							nc.skip = false
+							continue
+						}
+						nc.msgChan <- &nmqMessage{
+							raw: msg,
+						}
+					case <-nc.closeChan:
+						log.Info("close nmq consumer ", zap.String("topic", nc.topic), zap.String("groupName", nc.groupName))
+						close(nc.msgChan)
+						return
+					}
+				}
+			}()
+		})
+	}
+	return nc.msgChan
+}
+
+// Seek is used to seek the position in natsmq topic
+func (nc *Consumer) Seek(id mqwrapper.MessageID, inclusive bool) error {
+	if err := nc.closed(); err != nil {
+		return err
+	}
+	if nc.sub != nil {
+		return fmt.Errorf("can not seek() on an initilized consumer")
+	}
+	msgID := id.(*nmqID).messageID
+	// skip the first message when consume
+	nc.skip = !inclusive
+	// TODO: check the subscription config/policy.
+	var err error
+	nc.sub, err = nc.js.ChanSubscribe(nc.topic, nc.natsChan, nats.StartSequence(msgID))
+	return err
+}
+
+// Ack is used to ask a natsmq message
+func (nc *Consumer) Ack(message mqwrapper.Message) {
+	if err := message.(*nmqMessage).raw.Ack(); err != nil {
+		log.Warn("failed to ack message of nmq", zap.String("topic", message.Topic()), zap.Reflect("msgID", message.ID()))
+	}
+}
+
+// Close is used to free the resources of this consumer
+func (nc *Consumer) Close() {
+	if err := nc.sub.Unsubscribe(); err != nil {
+		log.Warn("failed to unsubscribe subscription of nmq", zap.String("topic", nc.topic))
+	}
+	close(nc.closeChan)
+}
+
+// GetLatestMsgID returns the ID of the most recent message processed by the consumer.
+func (nc *Consumer) GetLatestMsgID() (mqwrapper.MessageID, error) {
+	if err := nc.closed(); err != nil {
+		return nil, err
+	}
+
+	info, err := nc.js.StreamInfo(nc.topic)
+	if err != nil {
+		return nil, util.WrapError("failed to get stream info of nats jetstream", err)
+	}
+	msgID := info.State.LastSeq
+	return &nmqID{messageID: msgID}, nil
+}
+
+// CheckTopicValid verifies if the given topic is valid for this consumer.
+// A consumer is tied to a specific topic, and thus in a multi-tenant situation,
+// should only be used to check messages from its associated topic.
+func (nc *Consumer) CheckTopicValid(topic string) error {
```

Comment:
This API should be used for checking topic whether exists  in Nats

Reply:
I think it already does that.
The consumer is tied to a topic. So we only need to check "topic != nc.topic" (line 143).

Reply:
CheckTopicValid will not only check the topic with self for others MQ, the semantics seems to be different with the current implementation.

Reply:
CheckTopicValid check following two things:
1. topic should exist.
2. topic should be empty.

new commit changed:
CheckTopic keep the semantic same with kafka client.